lib/spack/spack/fetch_strategy.py

```diff
@@ -792,11 +796,585 @@ def __str__(self):
         return "[go] %s" % self.url
 
 
+@six.add_metaclass(abc.ABCMeta)
+class GitRef(object):
+    """A wrapper that knows how to format git refspecs in multiple contexts."""
+    known_types = ['Branch', 'Commit', 'Tag']  # type: ClassVar[List[str]]
+    ref_type = None                            # type: ClassVar[str]
+    _ref = None                                # type: str
+
+    Tag = None                # type: ClassVar[Type[GitTag]]
+    Commit = None             # type: ClassVar[Type[GitCommit]]
+    Branch = None             # type: ClassVar[Type[GitBranch]]
+
+    def __init__(self, ref):
+        # type: (str) -> None
+        if self.ref_type not in self.known_types:
+            # This is made a TypeError and not an InvalidGitRef because it cannot arise
+            # from invalid user input.
+            raise TypeError('GitRef can only have the types {0}, but was given {1!r}'
+                            .format(self.known_types, self.ref_type))
+
+        if not isinstance(ref, six.string_types):
+            # This case can arise from user error when editing package.py files, so we
+            # want to catch this and handle it specially to provide more context.
+            raise InvalidGitRef('git reference was not a string: {0!r}'.format(ref))
+        self._ref = ref
+
+    def __repr__(self):
+        return '{0}(ref_type={1!r}, ref={2!r})'.format(type(self).__name__,
+                                                       self.ref_type, self._ref)
+
+    def __str__(self):
+        return '<{0}>'.format(self.repo_info_for_reference())
+
+    @abc.abstractproperty
+    def should_validate_matches_hash(self):
+        # type: () -> Optional[str]
+        """Return a hash prefix that the repo checkout should be compared against."""
+
+    @classmethod
+    def from_version_directive(cls, kwargs):
+        # type: (Dict) -> GitRef
+        # (1) Extract the relevant arguments.
+        commit = kwargs.pop('commit', None)
+        tag = kwargs.pop('tag', None)
+        branch = kwargs.pop('branch', None)
+        version_name = kwargs.pop('version_name', None)
+        # (2) Ensure no mutually exclusive kwargs are provided.
+        num_specified = len(list(filter(None, [commit, tag, branch])))
+        # (3) If no explicit git ref arguments are provided, assume that the version
+        # string itself is to be used as the branch name.
+        if (num_specified == 0) and version_name is not None:
+            return cls.Branch(version_name)
+        # (4) If we are given a tag or a branch with a commit sha, note that by using
+        # a special constructor.
```

Comment:
@cosmicexplorer your comments are *chef's kiss* :kissing_heart:  I wish there were more developers like yourself!

Reply:
I don't want others to have to solve any of these problems after me so I make sure to write these things down!